treescript/src/treescript/merges.py

```diff
@@ -0,0 +1,180 @@
+        current_version = get_version("browser/config/version.txt", repo_path)
+        next_version = f"{current_version.major_number}.{current_version.minor_number}"
+
+        await do_bump_version(config, repo_path, merge_config["version_files"], next_version, dontbuild=True, commit=False)
+    if merge_config.get("version_files_suffix"):
+        current_version = get_version("browser/config/version.txt", repo_path)
+        next_version = f"{current_version.major_number}.{current_version.minor_number}{merge_config.get('version_suffix')}"
+
+        await do_bump_version(config, repo_path, merge_config["version_files_suffix"], next_version, dontbuild=True, commit=False)
+
+    for f in merge_config.get("copy_files", list()):
+        shutil.copyfile(os.path.join(repo_path, f["src"]), os.path.join(repo_path, f["dst"]))
+
+    for f, from_, to in merge_config.get("replacements", list()):
+        replace(os.path.join(repo_path, f), from_, to)
+
+    if merge_config.get("remove_locales"):
+        log.info("Removing locales")
+        remove_locales(os.path.join(repo_path, "browser/locales/shipped-locales"), merge_config["remove_locales"])
+    touch_clobber_file(repo_path)
+
+
+def remove_locales(file_name, removals):
+    """Remove locales from shipped-locales (mozilla-release only)."""
+    with open(file_name) as f:
+        contents = f.readlines()
+    new_contents = ""
+    for line in contents:
+        locale = line.split()[0]
+        if locale not in removals:
+            new_contents += line
+        else:
+            log.info("Removed locale: %s" % locale)
+    with open(file_name, "w") as f:
+        f.write(new_contents)
+
+
+# do_merge {{{1
+async def do_merge(config, task, repo_path):
+    """Perform a merge day operation.
+
+    This function takes its inputs from task by using the ``get_merge_flavor``
+    function from treescript.task. Using `merge_info` to determine which
+    set of modifications to make. The modifications are stored in the
+    scriptworker to provide version control.
+
+    Args:
+        config (dict): the running config
+        task (dict): the running task
+        repo_path (str): the source directory
+
+    Raises:
+        TaskverificationError: if a merge flavor is not recognised..
+
+    Returns:
+        list: A list of the branches that need pushing, and the corresponding revision.
+              This is unlike other actions as the list of outgoing changes is
+              not related to the number of commands we've performed, but we do need
+              to know which branches to push.
+    """
+    flavor = get_merge_flavor(task)
+
+    if flavor not in merge_configs:
+        raise TaskVerificationError("Unknown configuration for merge day flavor {}".format(flavor))
+
+    from_branch = merge_configs[flavor].get("from_branch")
+    to_branch = merge_configs[flavor].get("to_branch")
+
+    log.info("hg pull %s", repo_path)
+    await run_hg_command(config, "pull", repo_path=repo_path)
```

Comment:
The only repo we know about is mozilla-unified, so does it add anything?

Reply:
I generally prefer to be explicit when interacting with things in automation. 

`hg robustcheckout` doesn't set `hg config paths.default`, except on the initial clone. If some other action were to were to clone mozilla-central rather than mozilla-unified and that action happened to run on this worker first, this would do the wrong thing.

Additionally, the following code depends on all the mozilla-* bookmarks/firefoxtree heads being pulled, so I think being explicit here helps people figuring out what is going on. (Note that for thunderbird usages of treescript, the repo will be a clone of comm-central (since there is a comm-unified). That doesn't affect this code, but does mean that people looking at this code might not assume the repo has a `paths.default = .../mozilla-unified`).